lib/lfrfid/lfrfid_raw_worker.c

```diff
@@ -0,0 +1,602 @@
+                    ctx->tmp_data,
+                    ctx->tmp_data_length);
+                ctx->buffers[id].size += ctx->tmp_data_length;
+                ctx->tmp_data_length = 0;
+                ctx->buffers_counter = id;
+
+                overrun = false;
+            }
+        }
+
+        if(overrun) {
+            ctx->overrun_count++;
+        }
+    }
+
+    portYIELD_FROM_ISR(xHigherPriorityTaskWoken);
+}
+
+static int32_t lfrfid_raw_read_worker_thread(void* thread_context) {
+    LFRFIDRawWorker* worker = (LFRFIDRawWorker*)thread_context;
+
+    Storage* storage = furi_record_open("storage");
+    Stream* stream = file_stream_alloc(storage);
+    const char* filename = string_get_cstr(worker->file_path);
+    bool file_valid = file_stream_open(stream, filename, FSAM_READ_WRITE, FSOM_CREATE_ALWAYS);
+
+    LFRFIDRawWorkerReadData* data = malloc(sizeof(LFRFIDRawWorkerReadData));
+
+    data->ctx.overrun_count = 0;
+    data->ctx.stream =
+        xStreamBufferCreate(sizeof(RfidReadBuffer*) * READ_BUFFER_COUNT, sizeof(RfidReadBuffer*));
+
+    if(file_valid) {
+        // write header
+        size_t size = stream_write_cstring(stream, RAW_HEADER);
+        if(size != strlen(RAW_HEADER)) {
+            file_valid = false;
+        }
+
+        uint32_t max_buffer_size = RFID_DATA_BUFFER_SIZE;
+        size = stream_write(stream, (uint8_t*)&max_buffer_size, sizeof(uint32_t));
+        if(size != sizeof(uint32_t)) {
+            file_valid = false;
+        }
+
+        size = stream_write(stream, (uint8_t*)&worker->frequency, sizeof(float));
+        if(size != sizeof(float)) {
+            file_valid = false;
+        }
+
+        size = stream_write(stream, (uint8_t*)&worker->duty_cycle, sizeof(float));
+        if(size != sizeof(float)) {
+            file_valid = false;
+        }
+    }
+
+    if(file_valid) {
+        // setup carrier
+        furi_hal_rfid_pins_read();
+        furi_hal_rfid_tim_read(worker->frequency, worker->duty_cycle);
+        furi_hal_rfid_tim_read_start();
+
+        // stabilize detector
+        furi_delay_ms(1500);
+
+        // start capture
+        furi_hal_rfid_tim_read_capture_start(lfrfid_raw_worker_capture, &data->ctx);
+
+        RfidReadBuffer* buffer_p = NULL;
+        while(1) {
+            size_t size =
+                xStreamBufferReceive(data->ctx.stream, &buffer_p, sizeof(RfidReadBuffer*), 100);
+
+            if(size == sizeof(RfidReadBuffer*)) {
+                size_t size = 0;
+
+                size = stream_write(stream, (uint8_t*)&buffer_p->size, sizeof(size_t));
+                if(size != sizeof(size_t)) {
+                    file_valid = false;
+                }
+
+                size = stream_write(stream, buffer_p->data, buffer_p->size);
+                if(size != buffer_p->size) {
+                    file_valid = false;
+                }
+
+                buffer_p->size = 0;
+                __DMB();
+                buffer_p->occupied = false;
```

Comment:
volatile?

Reply:
volatile will load the cpu more than we can afford, this place is extremely sensitive to that